server_modules/scores.js

```diff
@@ -15,6 +16,32 @@ function reduceToMap(key) {
     }
 }
 
+function changeScores(action) {
+    var path = fileSystem.getDataFilePath('scores.json');
+    return new Promise(function(res, rej) {
+        lockfile.lock('scores.json.lock', { retries: 5, retryWait: 100 }, function (err) {
+            if(err) rej(err);
+
+            promise = fileSystem.readJsonFile(path)
+            .then(function(data) {
+                return data;
+            })
+            .catch(function() {
+                return { version:3, scores: [] };
+            })
+            .then(action)
+            .then(function(scores) {
+                fileSystem.writeFile(path, JSON.stringify(scores));
+
+                lockfile.unlock('scores.json.lock', function(err) {
+                    if(err) rej(err);
+                });
+                return scores;
+            }).then(res).catch(rej);
```

Comment:
This is problematic because of the nested functions of lockfile. I cannot return the promise chain, since it is being created inside the lockfile.lock function.

Reply:
You can, and you should.
Again, you have to think differently: the first lockfile-stuff should already do resolve/reject on the promise, then the next part can start, etc.

This is very tricky to get right though, so here's my attempt at it (warning: didn't run or compile this code myself, so please doublecheck!)

```js
/**
 * Create lockfile for given path.
 * @param path Path to lockfile (e.g. "something.json.lock")
 * @return Promise that resolves when lockfile is created
 */
function lockFile(path) {
    return new Promise((res, rej) => {
        lockfile.lock(path, { retries: 5, retryWait: 100 }, (err) => {
            if (err) {
                return rej(err);
            }
            res();
        });
    })
}

/**
 * Unlock lockfile for given path.
 * @param path Path to lockfile (e.g. "something.json.lock")
 * @return Promise that resolves when lockfile is unlocked.
 */
function unlockFile(path) {
    return new Promise((res, rej) => {
        lockfile.unlock(path, (err) => {
            if (err) {
                return rej(err);
            }
            res();
        });
    })
}

/**
 * Atomically change scores file.
 * Action callback is called with the current contents of the scores file, and is expected
 * to return the new contents (or a Promise for it).
 * A lock is acquired and held during the entire operation.
 * @param action Callback that receives current scores.json object, must return new contents (or Promise for it)
 * @return Promise for updated scores object
 */
function changeScores(action) {
    var scoresPath = fileSystem.getDataFilePath('scores.json');
    var lockfilePath = scoresPath + '.lock';
    var scores;
    return lockFile(lockfilePath)
        .then(() => fileSystem.readJsonFile(scoresPath))
        .catch(function (err) {
            if (err.status === 404) {
                // File doesn't exist, return initial version
                return { version: 3, scores: [] };
            }
            throw err;
        })
        .then(action)
        .then((scores) => {
            scores = scores;
            return fileSystem.writeFile(scoresPath, JSON.stringify(scores))
        })
        .then(
            () => unlockFile(lockfilePath),
            (err) => {
                // always unlock, also in case of errors, but make sure
                // the original error is still returned
                return unlockFile(lockfilePath)
                    .then(
                        () => { throw err; },
                        () => { throw err; }
                    );
            }
        )
        .then(() => scores);
}
```

I did quite a lot of things differently:
- lockfile is now created next to the file it's about, which makes it more robust when multiple processes (i.e. non-fllscoring) would use this mechanism.
- If an error happens in the more interesting parts (the action and/or writeFile), that error is indeed returned, instead of being swallowed by the unlock
- The file-not-found code checks for a more stable thing (message is typically a human-readable thing and should not be considered reliable)
- Pulled the lockfile stuff to its own functions to promisify them
- Added doc comments

That whole 'tail' with the rethrowing of errors isn't looking very nice, but it's a bit of a consequence of using native promises (that don't have handy helpers) and the fact that we're trying to 'bypass' the unlock, but still want to wait for its completion.

Reply:
I just now added server_modules/lock.js for that. Please tell me what you think